Explicitly add RTs to agent message format

sebageek opened the pull request:
We now put the Route Targets that should be exported/imported directly
into the agent message / SwitchConfig format. This means that we'll have
less business logic inside the agent and more in the neutron-server part
of the driver (which is a good thing, as we will have multiple agents
and therefore will not need to duplicate business logic). It will also
help when we implement the config diff functionality, as we can then
just fill the SwitchConfig object with what's on the switch and diff it
with the "should" version from OpenStack.

sebageek committed:
Explicitly add RTs to agent message format

We now put the Route Targets that should be exported/imported directly
into the agent message / SwitchConfig format. This means that we'll have
less business logic inside the agent and more in the neutron-server part
of the driver (which is a good thing, as we will have multiple agents
and therefore will not need to duplicate business logic). It will also
help when we implement the config diff functionality, as we can then
just fill the SwitchConfig object with what's on the switch and diff it
with the "should" version from OpenStack.

sebageek committed:
RPC caller tool now accepts kwargs as json

To better debug and test RPC calls we can now pass kwargs encoded as
json via the command line, e.g. --kwargs '{"switches": ["foo"]}'.

sebageek committed:
Add RPC call to get config from switches

We now can grab the config from a switch (or a number of switches). The
config will be put into a SwitchConfigUpdate object for transport and
vendor independency. This means that config snippets that are vendor
specific and/or are implicitly generated (for example the "redistribute
learned" flag in the EOS case) are not shown.

With this functionality we are one step closer to proper config diffing
capabilities.

sebageek committed:
Add get config to switches/switchgroups API

Now we can get the config of a switch or switchgroup via API and receive
the "already known" internal representation of the config.

sebageek committed:
Remove unnecessary print() from tests

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+    if rt.isnumeric():
+        rt = int(rt)
+        if (rt >> 48) & 0xFF != 0x2:
+            raise ValueError("Route targets are extended attributes and need to have a 0x2 in their low order position "
```
`and need to have a 0x2 in the type field's low order position` maybe?

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+        if rt_type == 0:
+            # 2 byte asn, 4byte number
+            a = (rt >> 32) & (2**16 - 1)
+            b = rt & (2**32 - 1)
```
Maybe call this `admin_field` in accordance with the RFC?

sebageek committed:
Remove old Vxlan maps on config update

When introducing new vni-vlan mappings on switch config update
(operation add) we need to look onto the device if there is already an
old mapping for either the vni or vlan id that we are referencing. These
need to be removed in order for the new mappings to be properly applied
to the switch. Sadly, this means that we have to actually do a lookup on
the switch each time we add one or more vni/vlan mappings, but that is
just the world we life in. On add operation we log a warning if there
are stale mappings, as this generally should not happen that often (only
when "things break" or devices are unreachable).

swagner-de commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -45,29 +93,49 @@ class VXLANMapping(pydantic.BaseModel):
+
+        rd = f"{self.asn}:{vni}"
         for bv in self.vlans:
-            if bv.rd == rd and bv.vlan == vlan and bv.vni == vni:
+            if bv.rd == rd and bv.vlan == vlan:
                 return
```
I am not sure if validators are run everytime an objects property is changed, but if so, couldn't that be a validator?
Also, is it a good idea to silently stop? Is it expected that we add VLANs to a BGP object that it is not responsible for?

sebageek replied:
The idea here is that if we construct the config update and due "some sort of codeflow" it might happen that we add the same vlan mapping twice. I want to make sure we don't end up in config generation with something that tries to add the same thing twice and then fails.

Validators are not run on object modification:
```
In [37]: agent_msg.BGPVlan(rd="123", vlan=1234, rt_imports=["123"])
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
Input In [37], in <module>
----> 1 agent_msg.BGPVlan(rd="123", vlan=1234, rt_imports=["123"])

File ~/venvs/fabric/lib/python3.8/site-packages/pydantic/main.py:331, in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for BGPVlan
rt_imports -> 0
  Route targets are extended attributes and need to have a 0x2 in the type field's low order position - 123 does not have this set (RFC 4360 sec 4) (type=value_error)

In [38]: bv = agent_msg.BGPVlan(rd="123", vlan=1234)

In [39]: bv.rt_imports.append("123")

In [40]: 
```
Though calling validation is possible (at least it reads like it), see here: https://github.com/samuelcolvin/pydantic/issues/1864

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+            raise ValueError(f"Rt {rt} has invalid type {rt_type}, see RFC 4364 sec 4.2")
+    else:
+        # ip:num, asn:num, num:num
+        m = re.match(r"^(?P<first>(?:\d+\d+\.\d+\.\d+)|(?:\d+\.\d+)|(?:\d+)):(?P<second>\d+)$", rt)
```
Why `\d+\d+`? Doesn't the first `\d+` already contain the second?

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+        # ip:num, asn:num, num:num
+        m = re.match(r"^(?P<first>(?:\d+\d+\.\d+\.\d+)|(?:\d+\.\d+)|(?:\d+)):(?P<second>\d+)$", rt)
+        if not m:
+            raise ValueError("RT is not in a recognizable format")
```
Do we maybe want to know which RT here?

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+        m = re.match(r"^(?P<first>(?:\d+\d+\.\d+\.\d+)|(?:\d+\.\d+)|(?:\d+)):(?P<second>\d+)$", rt)
+        if not m:
+            raise ValueError("RT is not in a recognizable format")
+        # automatically reformat cases where asn is > 16bit to as dot notation
```
`to as` -> `as` or `to` only?

sebageek replied:
`as`  to `AS`

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+            raise ValueError("RT is not in a recognizable format")
+        # automatically reformat cases where asn is > 16bit to as dot notation
+        first = m.group('first')
+        if first.isnumeric() and int(first) >= 2**16:
```
You might want `isdecimal()` instead.

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -26,6 +28,52 @@
+        first = m.group('first')
+        if first.isnumeric() and int(first) >= 2**16:
+            first = int(first)
+            rt = f"{first >> 16}.{first & (2**16 -1)}:{m.group('second')}"
```
is there a space missing before `1` and after `-`?

joker-at-work commented on networking_ccloud/tools/agent_rpc_caller.py:
```diff
@@ -58,10 +60,16 @@ def main():
+        print(f"Error: Could not decode json in kwargs: {e}")
+        sys.exit(1)
+
     print(f"Doing RPC call {CONF.method}(*{CONF.args}) with topic {topic}")
```
Does this not need the kwargs? Imho, it now shows incomplete information about the call that was being made.

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -211,13 +211,28 @@ def add_iface(self, iface):
                 return iface
+        return None
+
+    def get_or_create_iface(self, name):
+        iface = self.get_iface(name)
+        if iface:
+            return iface
+
+        iface = IfaceConfig(name=name)
+        self.ifaces.append(iface)
+        return iface
```
Could this be racy by being called from different threads/greenthreads?

sebageek replied:
These objects are generally created by a single thread and are not designed for two threads creating it at the same time. The thing would need a whole redesign for that.

joker-at-work commented on networking_ccloud/ml2/agent/common/messages.py:
```diff
@@ -211,13 +211,28 @@ def add_iface(self, iface):
-    def get_or_create_iface(self, switchport):
+    def get_iface(self, name):
         if self.ifaces is None:
             self.ifaces = []
```
Is order important or could this be a dict for faster access and for not having duplicates? Looks like `.name` is key anyways.

joker-at-work commented on networking_ccloud/ml2/agent/eos/switch.py:
```diff
@@ -236,6 +242,72 @@ def get_removable_translations_cmds(iface_name, wanted_translations):
+            config.add_vlan(int(vlan), data['name'])
+
+        # vxlan_maps
+        vxlan_maps = self.send_cmd("show interfaces Vxlan1")['interfaces']['Vxlan1']['vlanToVniMap']
```
`Vxlan1` is hardcoded? O.o

sebageek replied:
`Vxlan1` is a special interface on the `EOS` platform which contains all the vxlan mappings and some other magical stuff. In theory™ there could be multiple Vxlan interfaces, but no EOS platform supports that yet to my knowedge (and our usecase wouldn't need it).

joker-at-work commented on networking_ccloud/ml2/agent/eos/switch.py:
```diff
@@ -236,6 +242,72 @@ def get_removable_translations_cmds(iface_name, wanted_translations):
+        for vlan_name, data in curr_bgp_vlans.items():
+            m = vre.match(vlan_name)
+            if not m:
+                LOG.warning("Could not match bgp vpn instanec name '%s'", vlan_name)
```
`instanec` -> `instance`

joker-at-work commented on networking_ccloud/ml2/agent/eos/switch.py:
```diff
@@ -236,6 +242,72 @@ def get_removable_translations_cmds(iface_name, wanted_translations):
+            if name == "Vxlan1":
+                continue
+            iface = config.get_or_create_iface(name)
+            if data.get("untaggedVlan", 1) != 1:
```
Can this be `'1'`? We converted a lot to `int()` above, so I'm curious why this comes back as `int` already.

sebageek replied:
Values in the `EOS` json response are mostly an int where they should be an int. Dict keys on the other hand are strings, so they need int converting.

```
qa-de-1-sw4113a-bb271#show interfaces vlans | json | inc untagged
            "untaggedVlan": 2121
            "untaggedVlan": 2121
            "untaggedVlan": 2121
```

joker-at-work commented on networking_ccloud/ml2/agent/eos/switch.py:
```diff
@@ -113,13 +113,27 @@ def _make_config_from_update(self, config):
+                    # purge mappings referenced by config update, but point to something else
+                    for vlan, data in curr_maps.items():
+                        switch_vlan = int(vlan)
+                        switch_vni = data['vni']
```
We use `data.get('vni')` below, why are we sure it exists here?
